File: bestia-zoneserver/src/main/java/net/bestia/zoneserver/actor/battle/AttackUseActor.java
package net.bestia.zoneserver.actor.battle;

import java.util.Objects;

import net.bestia.messages.entity.EntitySkillUseMessage;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import net.bestia.messages.attack.AttackUseMessage;
import net.bestia.messages.entity.EntityDamageMessage;
import net.bestia.model.battle.Damage;
import net.bestia.zoneserver.actor.SpringExtension;
import net.bestia.zoneserver.actor.zone.ClientMessageHandlerActor.RedirectMessage;
import net.bestia.zoneserver.actor.zone.SendClientsInRangeActor;
import net.bestia.zoneserver.battle.BattleService;

/**
 * Transforms the incoming player attack message into a entity skill message
 * after some sanity checks.
 * 
 * @author Thomas Felix
 *
 */
@Component
@Scope("prototype")
public class AttackUseActor extends AbstractActor {

	public final static String NAME = "attackUse";

	private final LoggingAdapter LOG = Logging.getLogger(getContext().system(), this);

	private final BattleService battleService;
	private final ActorRef transformAtkMsg;
	private final ActorRef sendActiveRange;

	@Autowired
	public AttackUseActor(BattleService battleService, ActorRef msgHub) {

		this.battleService = Objects.requireNonNull(battleService);
		this.transformAtkMsg = SpringExtension.actorOf(getContext(), AttackPlayerUseActor.class);
		this.sendActiveRange = SpringExtension.actorOf(getContext(), SendClientsInRangeActor.class);
	}

	@Override
	public Receive createReceive() {
		return receiveBuilder()
				.match(AttackUseMessage.class, this::handleAttackMessage)
				.match(EntitySkillUseMessage.class, this::handleEntitySkillMessage)
				.build();
	}

	@Override
	public void preStart() throws Exception {
		final RedirectMessage msg = RedirectMessage.get(AttackUseMessage.class, EntitySkillUseMessage.class);
		context().parent().tell(msg, getSelf());
	}

	/**
	 * This message is received directly from the clients and must be translated
	 * first.
	 * 
	 * @param msg
	 */
	private void handleAttackMessage(AttackUseMessage msg) {
		LOG.debug("Received essage: {}.", msg);
		transformAtkMsg.tell(msg, getSelf());
	}

	/**
	 * Handles an attack by an entity to the ground or another entity.
	 * 
	 * @param msg
	 *            The message describing the attack.
	 */
	private void handleEntitySkillMessage(EntitySkillUseMessage msg) {
		LOG.debug("Received skill message: {}", msg);

		if (msg.getTargetEntityId() != 0) {
			// Entity was targeted.
			final Damage dmg = battleService.attackEntity(msg.getAttackId(), 
					msg.getSourceEntityId(),
					msg.getTargetEntityId());
			
			final EntityDamageMessage dmgMsg = new EntityDamageMessage(msg.getTargetEntityId(), dmg);
			sendActiveRange.tell(dmgMsg, getSelf());

		} else {
			LOG.warning("Attackmode Currently not supported.");
		}
	}
}

File: bestia-zoneserver/src/main/kotlin/net/bestia/zoneserver/actor/battle/AttackUseActor.kt
package net.bestia.zoneserver.actor.battle

import mu.KotlinLogging
import net.bestia.messages.entity.DamageType
import net.bestia.messages.entity.EntityDamage
import net.bestia.messages.entity.SkillUseEntity
import net.bestia.messages.entity.SkillUsePosition
import net.bestia.model.geometry.Vec3
import net.bestia.zoneserver.actor.Actor
import net.bestia.zoneserver.actor.SpringExtension
import net.bestia.zoneserver.actor.client.SendClientsInRangeActor
import net.bestia.zoneserver.actor.entity.awaitEntityResponse
import net.bestia.zoneserver.actor.routing.DynamicMessageRoutingActor
import net.bestia.zoneserver.actor.routing.MessageApi
import net.bestia.zoneserver.battle.BattleService
import net.bestia.zoneserver.entity.Entity
import net.bestia.zoneserver.entity.component.AttackListComponent

private val LOG = KotlinLogging.logger { }

/**
 * Transforms the incoming player attack message into a entity skill message
 * after some sanity checks.
 *
 * @author Thomas Felix
 */
@Actor
class AttackUseActor(
    private val battleService: BattleService,
    private val messageApi: MessageApi
) : DynamicMessageRoutingActor() {

  private val sendActiveRange = SpringExtension.actorOf(context, SendClientsInRangeActor::class.java)

  override fun createReceive(builder: BuilderFacade) {
    builder.matchRedirect(SkillUseEntity::class.java, this::handleAttackOnEntity)
    builder.matchRedirect(SkillUsePosition::class.java, this::handleAttackOnPosition)
  }

  /**
   * This message is received directly from the clients and must be translated
   * first.
   *
   * @param msg
   */
  private fun handleAttackOnEntity(msg: SkillUseEntity) {
    LOG.debug { "Use skill: $msg" }

    val attackerId = msg.entityId
    val defenderId = msg.targetEntityId

    awaitEntityResponse(messageApi, context, setOf(attackerId, defenderId)) {
      val attacker = it[attackerId]
      verifyAttackerKnowsAttack(attacker, msg.attackId)

      val dmgs = battleService.attackEntity(msg.attackId, it[attackerId], it[defenderId])

      dmgs.forEach { dmg ->
        val dmgMsg = EntityDamage(defenderId, dmg.amount, DamageType.DAMAGE)
        sendActiveRange.tell(dmgMsg, self)
      }
    }
  }

  /**
   * Handles an attack to the ground by calling the attack script.
   *
   * @param msg
   * The message describing the attack.
   */
  private fun handleAttackOnPosition(msg: SkillUsePosition) {
    LOG.trace { "handleAttackOnPosition: $msg" }

    awaitEntityResponse(messageApi, context, msg.entityId) { attacker ->
      verifyAttackerKnowsAttack(attacker, msg.attackId)

      val pos = Vec3(msg.x, msg.y, msg.z)
      battleService.attackGround(msg.attackId, attacker, pos)
    }
  }

  private fun verifyAttackerKnowsAttack(attacker: Entity, attackId: Long) {
    val knownAttackComp = attacker.getComponent(AttackListComponent::class.java)
    check(!knownAttackComp.knownAttacks.contains(attackId)) {
      "Attacker $attacker does not know attack ${attackId} only: $knownAttackComp"
    }
  }

  companion object {
    const val NAME = "attackUse"
  }
}
